design/005-Wire-Encoding-Abstraction.md

```diff
@@ -0,0 +1,108 @@
+# Proposal: Abstraction of Wire Messages Encoding
+
+* Author(s): Sebastian Stammler (@sebastianst)
+* Status: propose
+* Related issue: <!-- org-name/project#NNN, org-name/project#NNN. -->
+* Affected repository: go-perun
+
+## Summary
+
+The goal is to be able to choose different de/-serialization methods for wire
+messages when using go-perun's `wire` package for inter-node wire message
+exchange.
+
+## Motivation
+
+Wire messages in go-perun currently use a home-baked and non-exchangeable
+encoding and decoding format. When developing new Perun channel client
+implementations, it would currently be necessary to mirror this custom
+serialization format.
+
+## Details
+
+Wire messages in go-perun currently use a home-baked and non-exchangeable
+encoding and decoding format, as specified by `Encode` and `Decode`
+methods on wire messages (see, e.g., message definitions in `client/*msgs.go`),
+methods `Envelope.(En|De)code` and functions `(En|De)code` in `wire/msg.go`.
+
+It should be possible to choose a different wire serialization protocol, e.g., protobuf.
+
+### Note on `pkg/io` serialization
+
+Note, thought, that this custom format is _not_ Go-specific. It is a packed,
+little-endian binary encoding format and the serialization logic of actual Go
+types is defined in package `pkg/io` and in particular in file
+`pkg/io/serialize.go`. (However, I just realize that `*big.Int`s are actually
+encoded in big-endian byte order...so we might as well switch to big-endianess
+in `pkg/io/serialize.go` to be more consistent
+([related issue](https://github.com/hyperledger-labs/go-perun/issues/167)).)
+
+## Rationale
+
+See #motivation.
+
+## Impact
+
+This can be implemented in a backwards-compatible fashion, where `pkg/io`
+encoding would be the default and other serialization backends could be injected
+at runtime.
+
+## Implementation
+
+Luckily, this can be achieved with a few changes to package `wire` and the
+`wire.Msg` implementations itself.
+
+The core idea is to create an envelope serialization backend, which can be set
+at runtime (using `init()` of the respecive serialization backend
```

Comment:
We might want to not fix it globally, but set it at runtime, e.g., when initializing the message bus.

Reply:
We usually "activate" backends by passively importing them (see channel and wallet backend). Then the best place to put global initialization is in `init()` functions. Another reason why you should always follow this patter (use `init` if setting global stuff) is that the program will crash, that is, panic at startup, not during runtime. So I propose to leave it as a global `init` backend in the scope of this proposal.

Reply:
I think, in the first implementation of this abstraction, we could do it in the same way as in other backends (wallet, channel): having a global instance and activating them by passively importing.

As we [were discussing](https://github.com/hyperledger-labs/perun-proposals/pull/16#issuecomment-908211989), in the future, when we want to make the serializer instance local to a channel, we could set the backend in runtime (when initializing the channel).

What do you think ? @sebastianst @matthiasgeihs

Reply:
Yes, global passive per `init`, modular per channel/client/foo injected at runtime.

Reply:
added `e.g.` nonetheless ;)